nx/lib/nx/lin_alg.ex

```diff
@@ -967,4 +967,59 @@ defmodule Nx.LinAlg do
       opts
     )
   end
+
+  @doc """
+  Produces the tensor taken to the given power by dot-product.
+
+  The input is always a square tensor and a non-negative integer,
+  and the output is a square tensor of the same dimensions as the input tensor.
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 6)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [5743, 8370],
+          [12555, 18298]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.eye(3), 65535)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1]
+        ]
+      >
+
```

Comment:
We need an example for when power < 0.

I'd add one that shows that matrix_power(t, -1) == invert(t)

And another which has a bigger absolute value for the power.

Reply:
Btw, @josevalim, this makes me think that `**` should become matrix_power for when the left operand is a tensor. WDYT?

Reply:
The behaviour could change depending on what the right operator is: an integer could cause this function, and another equally-sized tensor could cause pairwise exponentiation.